src/base_obs.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2014 Swift Navigation Inc.
+ * Contact: Fergus Noble <fergus@swift-nav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <math.h>
+
+#include <libswiftnav/sbp_utils.h>
+#include <libswiftnav/pvt.h>
+#include <libswiftnav/constants.h>
+#include <libswiftnav/ephemeris.h>
+#include <libswiftnav/coord_system.h>
+#include <libswiftnav/linear_algebra.h>
+
+#include "board/leds.h"
+#include "position.h"
+#include "nmea.h"
+#include "sbp.h"
+#include "solution.h"
+#include "manage.h"
+#include "simulator.h"
+#include "settings.h"
+#include "timing.h"
+#include "base_obs.h"
+
+extern ephemeris_t es[MAX_SATS];
+
+/** \defgroup base_obs Base station observation handling
+ * \{ */
+
+/** Mutex to control access to the base station observations. */
+Mutex base_obs_lock;
+/** Semaphore that is flagged when a new set of observations are received. */
+BinarySemaphore base_obs_received;
+/** Most recent observations from the base station. */
+obss_t base_obss;
+
+/** Mutex to control access to the base station position state.
+ * (#base_pos_ecef and #base_pos_known) */
+Mutex base_pos_lock;
+/** Is the base station position known? i.e. is #base_pos_ecef valid? */
+bool_t base_pos_known = false;
+/** Base station known position in ECEF as sent to us by the base station in
+ * the BASE_POS message.  */
+double base_pos_ecef[3];
+
+/** SBP callback for when the base station sends us a message containing its
+ * known location.
+ * Stores the base station position in the global #base_pos_ecef variable and
+ * sets #base_pos_known to `true`.
+ */
+static void base_pos_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void) context; (void) len; (void) msg; (void) sender_id;
+
+  double llh_degrees[3];
+  double llh[3];
+  memcpy(llh_degrees, msg, 3*sizeof(double));
+
+  llh[0] = llh_degrees[0] * D2R;
+  llh[1] = llh_degrees[1] * D2R;
+  llh[2] = llh_degrees[2];
+
+  chMtxLock(&base_pos_lock);
+  wgsllh2ecef(llh, base_pos_ecef);
+  base_pos_known = true;
+  chMtxUnlock();
+}
+
+/** SBP callback for the old style observation messages.
+ * Just prints a deprecation warning. */
+static void obs_old_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
```

Comment:
You mean obs_old_callback? It would be hard to leave it functional but its nice to at least print a warning for a while. Its been this way for quite a while now.